aesara/tensor/math_opt.py

```diff
@@ -2283,37 +2284,32 @@ def local_log1p(fgraph, node):
 
 
 # TODO: in canonicalize, change log10 and log2 -> log
-@register_stabilize
-@register_specialize
 @local_optimizer([log])
-def local_log_add(fgraph, node):
-    # log(exp(x)+exp(y))
-    #
-    # Suppose x >= y
-    # log(exp(x) + exp(y))
-    # log(exp(x) * (1 + exp(y)/exp(x)))
-    # x + log(1 + exp(y)/exp(x))
-    # x + log1p(exp(y)/exp(x))
-    # x + log1p(exp(y-x))
+def local_log_add_exp(fgraph, node):
+    # log(exp(x)+exp(y)+exp(z)) -> max + log(x-max, y-max, z-max)
+
     if node.op == log:
         z = node.inputs[0]
         if z.owner and z.owner.op == add:
             zi = z.owner.inputs
-            if len(zi) != 2:
-                # -- upgrading Maximum to handle multiple inputs wasn't trivial
-                #    TODO
-                # raise NotImplementedError()
-                return
             pre_exp = [x.owner.inputs[0] for x in zi if x.owner and x.owner.op == exp]
+            # all arguments to add are exp(<something>)
             if len(pre_exp) == len(zi):
-                # all arguments to add are exp(<something>)
-                max_pre = maximum(*pre_exp)
-
-                ret = max_pre + log1p(exp(add(*[p - max_pre for p in pre_exp])))
-                ret.tag.values_eq_approx = values_eq_approx_remove_inf
+                max_pre = reduce(maximum, pre_exp)
+                # If max == -inf, offset would lead to nan output
+                max_pre = switch(isinf(max_pre), 0, max_pre)
+                ret = max_pre + log(add(*[exp(p - max_pre) for p in pre_exp]))
```

Comment:
Consider something like this:
```suggestion
                max_infs = isinf(max_pre)
                exp_sum_pre = switch(max_infs, 0, exp(add(*[p - max_pre for p in pre_exp])))
                ret = max_pre + log1p(exp_sum_pre)
```

Reply:
Can you move this `switch` into the `add` you were using?  That would break the `log(add(exp(...), ...))` pattern by producing `log(add(switch(..., 0, exp(...)), ...))`.

Reply:
Maybe. Will check it.

Reply:
This works, even though it feels a bit less elegant:
```python
max_pre = reduce(maximum, pre_exp)
ret = max_pre + log(add(*[
  switch(isinf(max_pre), exp(max_pre), exp(p - max_pre))
  for p in pre_exp
]))
```
Compared to the non-stable
```python
max_pre = reduce(maximum, pre_exp)
max_pre = switch(isinf(max_pre), 0, max_pre)
ret = max_pre + log(add(*[exp(p - max_pre) for p in pre_exp]))
```

What do you think?